File: examples/upstream.go
package main

import (
    "encoding/base64"
    "encoding/json"
    "net/http"
)

type ClientTLSInfo struct {
    Subject    string   `json:"subject"`
    URISANs    []string `json:"uri_sans"`
    DNSSANs    []string `json:"dns_sans"`
    Hash       string   `json:"hash"`
    NotBefore  string   `json:"not_before"`
    NotAfter   string   `json:"not_after"`
    Serial     string   `json:"serial"`
}

func index_handler(w http.ResponseWriter, r *http.Request) {
    if header := r.Header.Get("X-Client-TLS-Info"); header != "" {
        var info ClientTLSInfo
        decoded, _ := base64.StdEncoding.DecodeString(header)
        _ = json.Unmarshal(decoded, &info)
        _, _ = w.Write([]byte("Client Subject: " + info.Subject))
        return
    }
    w.WriteHeader(http.StatusUnauthorized)
}

func main() {
    server := &http.Server{}
    http.HandleFunc("/", index_handler)
    server.Addr = ":8080"
    _ = server.ListenAndServe()
}

File: examples/upstream.rs
use axum::http::StatusCode;
use axum::routing::get;
use axum::Router;
use base64::Engine;
use tokio::net::TcpListener;

#[derive(serde::Deserialize)]
struct ClientTLSInfo {
    subject: String,
    uri_sans: Vec<String>,
    dns_sans: Vec<String>,
    hash: String,
    not_before: String,
    not_after: String,
    serial: String,
}

async fn root_handler(req: axum::http::Request<axum::body::Body>) -> Result<String, StatusCode> {
    if let Some(h) = req.headers().get("X-Client-TLS-Info") {
        let b64_bytes = base64::engine::general_purpose::STANDARD
            .decode(h.to_str().unwrap())
            .unwrap_or_default();
        let json_str = String::from_utf8_lossy(&b64_bytes);
        let info: ClientTLSInfo = serde_json::from_str(&json_str).unwrap();
        Ok(format!("Client Subject: {}", info.subject))
    } else {
        Err(StatusCode::UNAUTHORIZED)
    }
}

#[tokio::main]
async fn main() {
    let router = Router::new()
        .route("/", get(root_handler));
    let listener = TcpListener::bind("0.0.0.0:8080").await.unwrap();
    tokio::spawn(async move {
        dbg!(axum::serve(listener, router).await.unwrap());
    });
}
